db/types.go

```diff
@@ -0,0 +1,198 @@
+
+// DBConnection represents a connection to a versioned database.
+// Records are accessed via transaction objects, and must be safe for concurrent creation
+// and read and write access.
+// Past versions are only accessible read-only.
+type DBConnection interface {
+	// Opens a read-only transaction at the current working version.
+	Reader() DBReader
+
+	// Opens a read-only transaction at a specified version.
+	// Returns ErrVersionDoesNotExist for invalid versions.
+	ReaderAt(uint64) (DBReader, error)
+
+	// Opens a read-write transaction at the current version.
+	ReadWriter() DBReadWriter
+
+	// Opens a write-only transaction at the current version.
+	Writer() DBWriter
+
+	// Returns all saved versions
+	Versions() (VersionSet, error)
+
+	// Saves the current contents of the database and returns the next version ID, which will be
+	// `Versions().Last()+1`.
+	// Returns an error if any open DBWriter transactions exist.
+	SaveNextVersion() (uint64, error)
+
+	// Attempts to save database at a specific version ID, which must be greater than or equal to
+	// what would be returned by `SaveVersion`.
+	// Returns an error if any open DBWriter transactions exist.
+	SaveVersion(uint64) error
+
+	// Deletes a saved version. Returns ErrVersionDoesNotExist for invalid versions.
+	DeleteVersion(uint64) error
+
+	// Close closes the database connection.
+	Close() error
+}
+
+// DBReader is a read-only transaction interface. It is safe for concurrent access.
+// Callers must call Discard when done with the transaction.
+//
+// Keys cannot be nil or empty, while values cannot be nil. Keys and values should be considered
+// read-only, both when returned and when given, and must be copied before they are modified.
+type DBReader interface {
+	// Get fetches the value of the given key, or nil if it does not exist.
+	// CONTRACT: key, value readonly []byte
+	Get([]byte) ([]byte, error)
+
+	// Has checks if a key exists.
+	// CONTRACT: key, value readonly []byte
+	Has(key []byte) (bool, error)
+
+	// Iterator returns an iterator over a domain of keys, in ascending order. The caller must call
+	// Close when done. End is exclusive, and start must be less than end. A nil start iterates
+	// from the first key, and a nil end iterates to the last key (inclusive). Empty keys are not
+	// valid.
+	// CONTRACT: No writes may happen within a domain while an iterator exists over it.
+	// CONTRACT: start, end readonly []byte
+	Iterator(start, end []byte) (Iterator, error)
+
+	// ReverseIterator returns an iterator over a domain of keys, in descending order. The caller
+	// must call Close when done. End is exclusive, and start must be less than end. A nil end
+	// iterates from the last key (inclusive), and a nil start iterates to the first key (inclusive).
+	// Empty keys are not valid.
+	// CONTRACT: No writes may happen within a domain while an iterator exists over it.
+	// CONTRACT: start, end readonly []byte
+	ReverseIterator(start, end []byte) (Iterator, error)
```

Comment:
my inclination is that you'd specify iteration direction as an option to the method that produes the iterator and not require duplicate methods.

Reply:
Yeah, that's a trivial change. Any other opinions on this?

Reply:
I don't have a strong preference, but tend to have the opposite inclination 🤷‍♂️

Reply:
keeping the interface minimal is a desired design. However it shouldn't on a cost of overloading the functions.

Reply:
I also don't have a strong opinion here, but we use different methods in the SDK currently. Being as we're only adding one parameter, I don't see why not.